File: solutions/elixir/056-subsets.ex
# Problem 56: Subsets (LeetCode 78)
# Difficulty: Med
# Language: Elixir
# 
defmodule Subsets do
  def subsets([]), do: [[]]
  def subsets([h | t]) do
    rest = subsets(t)
    rest ++ Enum.map(rest, &[h | &1])
  end
end

File: solutions/haskell/056-subsets.hs
{-
  Problem 56: Subsets (LeetCode 78)
  Difficulty: Med
  Language: Haskell
-}
subsets :: [a] -> [[a]]
subsets [] = [[]]
subsets (x:xs) =
  let rest = subsets xs
  in rest ++ map (x:) rest
